tools/linker/MonoTouch.Tuner/OptimizeGeneratedCodeSubStep.cs

```diff
@@ -68,198 +69,187 @@ protected override void Process (TypeDefinition type)
 
 		protected override void Process (MethodDefinition method)
 		{
-			// special processing on generated methods from NSObject-inherited types
-			// it would be too risky to apply on user-generated code
-			if (!method.HasBody || !method.IsGeneratedCode (LinkContext) || (!IsExtensionType && !IsExport (method)))
+			if (!method.HasBody)
 				return;
+
+			if (method.IsGeneratedCode (LinkContext) && (IsExtensionType || IsExport (method))) {
+				// We optimize methods that have the [GeneratedCodeAttribute] and is either an extension type or an exported method
+			} else {
+				// but it would be too risky to apply on user-generated code
+				return;
+			}
 			
 			var instructions = method.Body.Instructions;
 			for (int i = 0; i < instructions.Count; i++) {
-				switch (instructions [i].OpCode.Code) {
+				var ins = instructions [i];
+				switch (ins.OpCode.Code) {
 				case Code.Call:
-					ProcessCalls (method, i);
+					ProcessCalls (method, ins);
 					break;
 				case Code.Ldsfld:
-					ProcessLoadStaticField (method, i);
+					ProcessLoadStaticField (method, ins);
 					break;
 				}
 			}
 		}
 		
-		void ProcessCalls (MethodDefinition caller, int i)
+		void ProcessCalls (MethodDefinition caller, Instruction ins)
 		{
-			var instructions = caller.Body.Instructions;
-			Instruction ins = instructions [i];
 			var mr = ins.Operand as MethodReference;
-			// if it could not be resolved to a definition then it won't be NSObject
-			if (mr == null)
-				return;
-
-			switch (mr.Name) {
+			switch (mr?.Name) {
 			case "IsNewRefcountEnabled":
-				// note: calling IsNSObject would check inheritance (time consuming)
-				if (!mr.DeclaringType.Is (Namespaces.Foundation, "NSObject"))
-					return;
-
-				Nop (ins);							// call bool MonoTouch.Foundation.NSObject::IsNewRefcountEnabled()
-				ins = instructions [++i];			// brtrue IL_x
-				while (ins.OpCode.FlowControl != FlowControl.Cond_Branch)
-					ins = instructions [++i];		// csc debug IL is quite not optimal as can include an _unneeded_ stloc/ldloc[.x] (ref: #32282)
-				Instruction branch_to = (ins.Operand as Instruction);
-				while (ins != branch_to) {
-					//Console.WriteLine ("\t\t{0}", ins.OpCode.Code);
-					Nop (ins);						// for getters: ldarg.0 + ldloc.0 + stfld
-					ins = instructions [++i];		// for setters: ldarg.0 + ldarg.1 + stfld
-				}
+				ProcessIsNewRefcountEnabled (caller, ins);
```

Comment:
that's always enabled now (was optional only for classic and not even in the last classic release), so that can be removed.